RefugeeLand.Core.Api.Tests.Unit/Foundations/RefugeeServiceTests.Exceptions.Add.cs

```diff
@@ -0,0 +1,188 @@
+                new FailedRefugeeStorageException(sqlException);
+
+            var expectedRefugeeDependencyException =
+                new RefugeeDependencyException(failedRefugeeStorageException);
+
+            this.dateTimeBrokerMock.Setup(broker =>
+                    broker.GetCurrentDateTimeOffset())
+                .Throws(sqlException);
+
+            // when
+            ValueTask<Refugee> addRefugeeTask =
+                this.refugeeService.AddRefugeeAsync(someRefugee);
+
+            // then
+            await Assert.ThrowsAsync<RefugeeDependencyException>(() =>
+                addRefugeeTask.AsTask());
+
+            this.loggingBrokerMock.Verify(broker =>
+                    broker.LogCritical(It.Is(SameExceptionAs(
+                        expectedRefugeeDependencyException))),
+                Times.Once);
+
+            this.dateTimeBrokerMock.Verify(broker =>
+                    broker.GetCurrentDateTimeOffset(),
+                Times.Once);
+
+            this.storageBrokerMock.Verify(broker =>
+                    broker.InsertRefugeeAsync(It.IsAny<Refugee>()),
+                Times.Never);
+
+            this.loggingBrokerMock.VerifyNoOtherCalls();
+            this.storageBrokerMock.VerifyNoOtherCalls();
+            this.dateTimeBrokerMock.VerifyNoOtherCalls();
+        }
+
+        [Fact]
+        public async Task ShouldThrowDependencyValidationExceptionOnAddIfRefugeeAlreadyExistAndLogItAsync()
+        {
+            //given
+            Refugee randomRefugee = CreateRandomRefugee();
+            Refugee alreadyExistRefugee = randomRefugee;
+            string message = GetRandomString();
+            var duplicateKeyException = new DuplicateKeyException(message);
+            var alreadyExistException = new AlreadyExistRefugeeException(duplicateKeyException);
+
+            var expectedRefugeeDependencyValidationException =
+                new RefugeeDependencyValidationException(alreadyExistException);
+
+            this.dateTimeBrokerMock.Setup(broker =>
+                    broker.GetCurrentDateTimeOffset())
+                .Throws(duplicateKeyException);
+
+            //when
+            ValueTask<Refugee> addRefugeeTask =
+                this.refugeeService.AddRefugeeAsync(alreadyExistRefugee);
+
+            //then
+            await Assert.ThrowsAsync<RefugeeDependencyValidationException>(() =>
+                addRefugeeTask.AsTask());
+
+            this.loggingBrokerMock.Verify(broker =>
+                    broker.LogError(It.Is(SameExceptionAs(
+                        expectedRefugeeDependencyValidationException))),
+                Times.Once);
+
+            this.storageBrokerMock.Verify(broker =>
+                    broker.InsertRefugeeAsync(It.IsAny<Refugee>()),
+                Times.Never);
+
+            this.dateTimeBrokerMock.Verify(broker =>
+                    broker.GetCurrentDateTimeOffset(),
+                Times.Once);
```

Comment:
The business logic calls `dateTimeBroker` to throw exception, then `loggingBroker` is called. So this should be the order of `.Verify()` statements. The same applies to `VerifyNoOtherCalls()`